pkg/metrics/metrics.go

```diff
@@ -0,0 +1,190 @@
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/apimachinery/pkg/types"
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog"
+)
+
+const (
+	labelDriverName               = "driver_name"
+	labelOperationName            = "operation_name"
+	labelOperationStatus          = "operation_status"
+	snapshotController            = "snapshot_controller"
+	operationLatencyMetricName    = "operation_total_seconds"
+	operationLatencyMetricHelpMsg = "Total number of seconds spent by the snapshot controller for an operation to complete end to end"
+)
+
+type OperationState string
+
+const (
+	// InterimFailure is an operation state which means the controller will keep
+	// on retrying the operation upon encountering this type of error.
+	InterimFailure OperationState = "InterimFailure"
+
+	// TerminatingFailure is an operation state which means the controller has encountered
+	// an error which it's not able to resume the operation and has marked the
+	// permanently failure of the operation.
+	TerminatingFailure OperationState = "TerminatingFailure"
+
+	// Success states that the controller has successfully executed the operation.
+	Success OperationState = "Success"
+)
+
+var operationLatencyMetricBuckets = []float64{0.1, 0.25, 0.5, 1, 2.5, 5, 10, 15, 30, 60, 120, 300, 600}
+
+type MetricsManager interface {
+	// StartServing starts the metrics endpoint at the specified addr/pattern for
+	// metrics managed by this MetricsManager. It spawn a goroutine to listen to
+	// and serve HTTP requests received on addr/pattern
+	// if the "pattern" is empty (i.e., ""), no endpoint will be started. An error
+	// will be returned if there is any.
+	StartServing(pattern, addr string, logger promhttp.Logger) error
+
+	// OperationStart takes in an operation and cache it's start time.
+	// if the operation already exists, it's an no-op.
+	OperationStart(op Operation)
+
+	// RecordMetrics records a metric point. Note that it will be an no-op if an
+	// operation has not been marked "Started" previously via invoking "OperationStart".
+	// op - the Operation which the metric is associated with.
+	// state - metrics manager operations differently to the cache based on the state
+	//         of the Operation:
+	//   InterimFailure: record a failure metric however keep operation's start time in cache
+	//   TerminatingFailure: record a failure metric and remove operation's start time from cache
+	//   Success: record a success metric and remove the operation's start time from cache
+	RecordMetrics(op Operation, state OperationState)
+}
+
+// Operation is a structure which holds information to identifier a snapshot
+// related operation
+type Operation struct {
+	// the name of the operation, for example: "CreateSnapshot", "DeleteSnapshot"
+	Name string
+	// the name of the driver which executes the operation
+	Driver string
+	// the resource UID to which the operation has been executed against
```

Comment:
In this case, it's the VolumeSnapshot UID?

Reply:
for this specific case yes.